File: web_portal/backend/supabase_client.py
"""
Supabase Client for Flask Backend
"""

from typing import Dict, Optional, List
from supabase import create_client, Client
from flask import current_app


class SupabaseFlaskClient:
    """Supabase client wrapper for Flask"""
    
    def __init__(self):
        """Initialize client"""
        self.client: Optional[Client] = None
    
    def init_app(self, app):
        """
        Initialize with Flask app
        
        Args:
            app: Flask application instance
        """
        url = app.config.get('SUPABASE_URL')
        key = app.config.get('SUPABASE_KEY')
        
        if not url or not key:
            raise ValueError("SUPABASE_URL and SUPABASE_KEY must be configured")
        
        self.client = create_client(url, key)
        app.supabase = self
    
    def verify_user_token(self, token: str) -> Optional[Dict]:
        """
        Verify user JWT token
        
        Args:
            token: JWT token from Authorization header
            
        Returns:
            User data if valid, None otherwise
        """
        try:
            user = self.client.auth.get_user(token)
            return user.user if user else None
        except Exception as e:
            current_app.logger.error(f"Token verification error: {e}")
            return None
    
    def get_certificate_by_id(self, cert_id: str) -> Optional[Dict]:
        """
        Get certificate by ID
        
        Args:
            cert_id: Certificate ID
            
        Returns:
            Certificate record if found
        """
        try:
            response = self.client.table('certificates')\
                .select('*')\
                .eq('cert_id', cert_id)\
                .single()\
                .execute()
            
            return response.data if response.data else None
            
        except Exception as e:
            current_app.logger.error(f"Database query error: {e}")
            return None
    
    def get_user_certificates(self, user_id: str, limit: int = 50, offset: int = 0) -> List[Dict]:
        """
        Get certificates for a user
        
        Args:
            user_id: User ID
            limit: Maximum number of records
            offset: Offset for pagination
            
        Returns:
            List of certificate records
        """
        try:
            response = self.client.table('certificates')\
                .select('*')\
                .eq('user_id', user_id)\
                .order('created_at', desc=True)\
                .limit(limit)\
                .offset(offset)\
                .execute()
            
            return response.data if response.data else []
            
        except Exception as e:
            current_app.logger.error(f"Database query error: {e}")
            return []
    
    def insert_verification_log(self, cert_id: str, result: str, 
                               ip_address: str, user_agent: str) -> bool:
        """
        Log verification attempt
        
        Args:
            cert_id: Certificate ID
            result: Verification result
            ip_address: Client IP
            user_agent: Client user agent
            
        Returns:
            True if successful
        """
        try:
            self.client.table('verification_logs').insert({
                'cert_id': cert_id,
                'verification_result': result,
                'ip_address': ip_address,
                'user_agent': user_agent
            }).execute()
            return True
        except Exception as e:
            current_app.logger.error(f"Log insert error: {e}")
            return False
    
    def get_certificate_file_url(self, user_id: str, cert_id: str, file_type: str = 'pdf') -> Optional[str]:
        """
        Get signed URL for certificate file
        
        Args:
            user_id: User ID
            cert_id: Certificate ID
            file_type: File type ('pdf' or 'json')
            
        Returns:
            Signed URL if successful
        """
        try:
            bucket_name = current_app.config.get('CERTIFICATE_BUCKET', 'certificates')
            file_path = f"{user_id}/{cert_id}.{file_type}"
            
            # Get signed URL (valid for 1 hour)
            response = self.client.storage.from_(bucket_name)\
                .create_signed_url(file_path, 3600)
            
            return response.get('signedURL') if response else None
            
        except Exception as e:
            current_app.logger.error(f"Storage URL error: {e}")
            return None
    
    def update_certificate_status(self, cert_id: str, status: str) -> bool:
        """
        Update certificate status
        
        Args:
            cert_id: Certificate ID
            status: New status
            
        Returns:
            True if successful
        """
        try:
            self.client.table('certificates')\
                .update({'status': status})\
                .eq('cert_id', cert_id)\
                .execute()
            return True
        except Exception as e:
            current_app.logger.error(f"Update error: {e}")
            return False
    
    def signup_with_profile(self, email: str, password: str, full_name: str) -> Dict:
        """
        Sign up user and create profile using service key
        
        Args:
            email: User email
            password: User password
            full_name: User full name
            
        Returns:
            Dict with user data and success status
        """
        try:
            # Create service client for admin operations
            service_key = current_app.config.get('SUPABASE_SERVICE_KEY')
            if not service_key:
                raise ValueError("SUPABASE_SERVICE_KEY not configured")
            
            service_client = create_client(
                current_app.config['SUPABASE_URL'], 
                service_key
            )
            
            # Sign up user
            auth_response = service_client.auth.sign_up({
                'email': email,
                'password': password,
                'options': {
                    'data': {
                        'full_name': full_name,
                    }
                }
            })
            
            if auth_response.user:
                # Create user profile
                profile_data = {
                    'id': auth_response.user.id,
                    'full_name': full_name
                }
                
                service_client.table('user_profiles').insert(profile_data).execute()
                
                return {
                    'success': True,
                    'user': {
                        'id': auth_response.user.id,
                        'email': auth_response.user.email,
                        'created_at': auth_response.user.created_at
                    },
                    'message': 'User created successfully. Please check your email to verify your account.'
                }
            else:
                return {
                    'success': False,
                    'error': 'Failed to create user'
                }
                
        except Exception as e:
            current_app.logger.error(f"Signup error: {e}")
            return {
                'success': False,
                'error': str(e)
            }

File: web_portal/backend/lib/supabaseClient.js
// web_portal/backend/lib/supabaseClient.js
import { createClient } from '@supabase/supabase-js';

class SupabaseClient {
  constructor() {
    const supabaseUrl = process.env.SUPABASE_URL;
    const supabaseKey = process.env.SUPABASE_KEY;
    const supabaseServiceKey = process.env.SUPABASE_SERVICE_KEY;

    if (!supabaseUrl || !supabaseKey) {
      throw new Error('SUPABASE_URL and SUPABASE_KEY must be configured');
    }

    this.client = createClient(supabaseUrl, supabaseKey);

    if (supabaseServiceKey) {
      this.serviceClient = createClient(supabaseUrl, supabaseServiceKey);
    }
  }

  async verifyUserToken(token) {
    try {
      const { data: { user }, error } = await this.client.auth.getUser(token);
      
      if (error) {
        console.error('Token verification error:', error);
        return null;
      }
      
      return user;
    } catch (error) {
      console.error('Token verification error:', error);
      return null;
    }
  }

  async getCertificateById(certId) {
    try {
      const { data, error } = await this.client
        .from('certificates')
        .select('*')
        .eq('cert_id', certId)
        .limit(1)
        .maybeSingle();

      if (error) {
        console.error('Database query error:', error);
        return null;
      }

      return data;
    } catch (error) {
      console.error('Database query error:', error);
      return null;
    }
  }

  async getUserCertificates(userId, limit = 50, offset = 0) {
    try {
      const { data, error } = await this.client
        .from('certificates')
        .select('*')
        .eq('user_id', userId)
        .order('created_at', { ascending: false })
        .range(offset, offset + limit - 1);

      if (error) {
        console.error('Database query error:', error);
        return [];
      }

      return data || [];
    } catch (error) {
      console.error('Database query error:', error);
      return [];
    }
  }

  async insertVerificationLog(certId, result, ipAddress, userAgent) {
    try {
      const { error } = await this.client
        .from('verification_logs')
        .insert({
          cert_id: certId,
          verification_result: result,
          ip_address: ipAddress,
          user_agent: userAgent
        });

      if (error) {
        console.error('Log insert error:', error);
        return false;
      }

      return true;
    } catch (error) {
      console.error('Log insert error:', error);
      return false;
    }
  }

  async getCertificateFileUrl(userId, certId, fileType = 'pdf', bucketName = 'certificates') {
    try {
      const filePath = `${userId}/${certId}.${fileType}`;
      
      const { data, error } = await this.client.storage
        .from(bucketName)
        .createSignedUrl(filePath, 3600); // Valid for 1 hour

      if (error) {
        console.error('Storage URL error:', error);
        return null;
      }

      return data?.signedUrl || null;
    } catch (error) {
      console.error('Storage URL error:', error);
      return null;
    }
  }

  async updateCertificateStatus(certId, status) {
    try {
      const { error } = await this.client
        .from('certificates')
        .update({ status })
        .eq('cert_id', certId);

      if (error) {
        console.error('Update error:', error);
        return false;
      }

      return true;
    } catch (error) {
      console.error('Update error:', error);
      return false;
    }
  }

  async signupWithProfile(email, password, fullName) {
    try {
      if (!this.serviceClient) {
        throw new Error('SUPABASE_SERVICE_KEY not configured');
      }

      // Sign up user
      const { data: authData, error: authError } = await this.serviceClient.auth.signUp({
        email,
        password,
        options: {
          data: {
            full_name: fullName,
          }
        }
      });

      if (authError) throw authError;

      if (authData.user) {
        // Create user profile
        const { error: profileError } = await this.serviceClient
          .from('user_profiles')
          .insert({
            id: authData.user.id,
            full_name: fullName
          });

        if (profileError) {
          console.error('Profile creation error:', profileError);
        }

        return {
          success: true,
          user: {
            id: authData.user.id,
            email: authData.user.email,
            created_at: authData.user.created_at
          },
          message: 'User created successfully. Please check your email to verify your account.'
        };
      }

      return {
        success: false,
        error: 'Failed to create user'
      };
    } catch (error) {
      console.error('Signup error:', error);
      return {
        success: false,
        error: error.message || 'Signup failed'
      };
    }
  }
}

// Create singleton instance
const supabaseClient = new SupabaseClient();

export default supabaseClient;